### ISSUE TITLE
option to treat `ALL_CAPS` variables as `Final`

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3650]
pyright's `reportConstantRedefinition` rule treats uppercase variables as `Final`:

```py
FOO = 1
FOO = 2 # error: "FOO" is constant (because it is uppercase) and cannot be redefined  (reportConstantRedefinition)
```

perhaps ty should support this too. alternatively, there could be a ruff rule that warns the user that uppercase variables do not behave as constants and to use `Final` instead.

**Comment by MichaReiser** [comment]
We could also make the rule so that it allows re-assignment if the variable has a non-final type annotation

**Comment by DetachHead** [comment]
you mean like this?

```py
FOO: int = 1

FOO: str = ""
```

i wouldn't expect that to be allowed (but i guess that would be covered by #2987?)

**Comment by MichaReiser** [comment]
Yeah. It feels very explicit to me to say it's not `Final`. But I don't feel strongly about it. I'd also be fine to disallow all

**Comment by carljm** [comment]
I like the ruff rule idea better, to alert people "hey it looks like maybe this should be Final". I'm not convinced that casing should have semantic effects in your type checker.

**Comment by carljm** [comment]
(I would be inclined to transfer this issue to Ruff instead.)
